components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -1925,6 +1928,137 @@ void JsonRpcService::GetERC1155TokenBalance(
       network_url, std::move(internal_callback));
 }
 
+void JsonRpcService::DiscoverAssets(
+    const std::string& chain_id,
+    const std::vector<std::string>&
+        account_addresses,  // TODO(nvonpentz) - Use AccountInfoPtr instead
+    std::vector<mojom::BlockchainTokenPtr> user_assets,
+    DiscoverAssetsCallback callback) {
+  if (chain_id != mojom::kMainnetChainId) {
+    std::move(callback).Run(
+        std::vector<mojom::BlockchainTokenPtr>(),
+        mojom::ProviderError::kMethodNotSupported,
+        l10n_util::GetStringUTF8(IDS_WALLET_METHOD_NOT_SUPPORTED_ERROR));
+    return;
+  }
+
+  if (account_addresses.empty()) {
+    std::move(callback).Run(
+        std::vector<mojom::BlockchainTokenPtr>(),
+        mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  for (const auto& account_address : account_addresses) {
+    if (!EthAddress::IsValidAddress(account_address)) {
+      std::move(callback).Run(
+          std::vector<mojom::BlockchainTokenPtr>(),
+          mojom::ProviderError::kInvalidParams,
+          l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+      return;
+    }
+  }
+
+  auto internal_callback = base::BindOnce(
+      &JsonRpcService::OnGetAllTokensDiscoverAssets,
+      weak_ptr_factory_.GetWeakPtr(), chain_id, account_addresses,
+      std::move(user_assets), std::move(callback));
+
+  BlockchainRegistry::GetInstance()->GetAllTokens(
+      chain_id, mojom::CoinType::ETH, std::move(internal_callback));
+  return;
+}
+
+void JsonRpcService::OnGetAllTokensDiscoverAssets(
+    const std::string& chain_id,
+    const std::vector<std::string>& account_addresses,
+    std::vector<mojom::BlockchainTokenPtr> user_assets,
+    DiscoverAssetsCallback callback,
+    std::vector<mojom::BlockchainTokenPtr> token_registry) {
+  auto network_url = GetNetworkURL(prefs_, chain_id, mojom::CoinType::ETH);
+  if (!network_url.is_valid()) {
+    std::move(callback).Run(
+        std::vector<mojom::BlockchainTokenPtr>(),
+        mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  base::Value topics(base::Value::Type::LIST);
+  if (!MakeAssetDiscoveryTopics(account_addresses, &topics)) {
+    std::move(callback).Run(
+        std::vector<mojom::BlockchainTokenPtr>(),
+        mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  // Create list of contract addresses to search by
+  // removing non erc20 tokens and assets that user has already
```

Comment:
probably this should have a privacy review because we are revealing some info to the network provider (infura?) and any proxies this is going through, and i'm not sure if this is all info that they can figure out already, like:

* what assets a user has already - this is probably always already known to them? unless there's assets on private blockchains that we support
* user's account addresses

Reply:
Filed https://github.com/brave/security/issues/1003